runtime/common/vec/vec_test.go

```diff
@@ -0,0 +1,333 @@
+package vec
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func wrong[T any](T) bool {
+	return false
+}
+
+func TestNewVecIsNil(t *testing.T) {
+	t.Parallel()
+	require.Nil(t, New[int]())
+
+	three := New(1, 2, 3)
+	require.Equal(t, []int(three), []int{1, 2, 3})
+}
+
+func TestVecPush(t *testing.T) {
+	t.Parallel()
+	vec := New[int]()
+	vec.Push(1)
+	require.Equal(t, []int(vec), []int{1})
+	vec.Push(2)
+	require.Equal(t, []int(vec), []int{1, 2})
+}
+
+func TestVecZeroed(t *testing.T) {
+	t.Parallel()
+	vec := Zeroed[int](5)
+	require.Equal(t, []int(vec), []int{0, 0, 0, 0, 0})
+}
+
+func TestVecPop(t *testing.T) {
+	t.Parallel()
+	t.Run("pop should remove from end", func(t *testing.T) {
+
+		t.Parallel()
+		v := Vec[int]([]int{1, 2, 3, 4, 5})
+		stack := make([]int, 0, 5)
+		for i := 0; i < 5; i++ {
+			val, ok := v.Pop()
+			require.True(t, ok)
+			stack = append(stack, val)
+		}
+	})
+
+	t.Run("pop should fail on empty vec", func(t *testing.T) {
+		t.Parallel()
+
+		vec := New[int]()
+		zero, ok := vec.Pop()
+		require.Zero(t, zero)
+		require.False(t, ok)
+	})
+}
+
+func TestVecConcat(t *testing.T) {
+	t.Parallel()
+	xs := New(1, 2, 3)
+	ys := New(4, 5, 6)
+	xs.Concat(ys)
+	require.Equal(t, xs, New(1, 2, 3, 4, 5, 6))
+}
+
+func TestVecFilter(t *testing.T) {
+	t.Parallel()
+
+	xs := New(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)
+
+	evens := xs.Filter(func(x int) bool {
+		return x%2 == 0
+	})
+
+	require.Equal(t, evens, New(2, 4, 6, 8, 10))
+
+	odds := xs.Filter(func(x int) bool {
+		return x%2 == 1
+	})
+
+	require.Equal(t, odds, New(1, 3, 5, 7, 9))
+
+	none := xs.Filter(wrong[int])
+	require.Empty(t, none)
+}
+
+func TestVecFind(t *testing.T) {
+	t.Parallel()
+
+	xs := New(0, 1, 2, 3, 4)
+	for i := 0; i < len(xs); i++ {
+		val, ok := xs.Find(func(x int) bool {
+			return x == i
+		})
+		require.True(t, ok)
+		require.Equal(t, i, val)
+	}
+	_, ok := xs.Find(wrong[int])
+
+	require.False(t, ok)
+}
+
+func TestVecMap(t *testing.T) {
+	t.Parallel()
+	xs := New(1, 2, 3)
+	ys := Map(xs, func(x int) string {
+		return fmt.Sprint(x)
+	})
+
+	// should be pure
+	require.Equal(t, New(1, 2, 3), xs)
+	require.Equal(t, New("1", "2", "3"), ys)
+}
+
+func TestVecFlatMap(t *testing.T) {
```

Comment:
How is the performance of these operations compared to using slices? Go generics tend to add some significant overhead. e.g: https://github.com/onflow/cadence/pull/2181#issuecomment-1361824071. 

Might be useful to add some benchmarks too.

Reply:
ugh i didn't even realize generics came with performance overhead. that tracks given the lack of type erasure, I'll write some benchmarks to investigate

Reply:
https://planetscale.com/blog/generics-can-make-your-go-code-slower

Interesting, seems like most of the generics overhead actually comes from vtable passing and indirections. The only code that uses parametric polymorphism are the numeric folds, `Sum()` and `Product()`, whereas the other methods are built on top of slice operations. This might hurt inlining but i'll post an update after benchmarking

Reply:
Ok, results aren't promising. Here are the results:

```
BenchmarkVecAll
BenchmarkVecAll/array_any
BenchmarkVecAll/array_any-8             1000000000               0.0000342 ns/op               0 B/op          0 allocs/op
BenchmarkVecAll/vec_any
BenchmarkVecAll/vec_any-8               1000000000               0.0002208 ns/op               0 B/op          0 allocs/op
```

given the following benchmark

```go

func BenchmarkVecAll(b *testing.B) {
	size := 100000
	array := make([]int, size + 1)
	array[size] = 1
	vec := Vec[int](array)
	for i := 0; i < b.N; i++ {

		b.Run("array any", func(b *testing.B) {
			res := false
			for _, n := range array {
				if n != 0 {
					res = true
					break
				}
			}
			require.True(b, res)
		})

		b.Run("vec any", func(b *testing.B) {
			res := vec.Any(func(x int) bool {
				return x != 0
			})
			require.True(b, res)
		})

	}
}
```

It seems that functional patterns incur a ton of overhead due to a lack of inlining at the call site, I'm not sure how much generics factor into this. But the almost 10x slowdown is alarming

Reply:
@dreamsmasher this shouldn't be this big

Reply:
```
goos: darwin
goarch: arm64
pkg: vextest/vec
BenchmarkVecAll/array_any-10         	1000000000	         0.003678 ns/op
BenchmarkVecAll/vec_any-10           	1000000000	         0.003284 ns/op
PASS
ok  	vextest/vec	0.166s
```

Reply:
```
goos: linux
goarch: amd64
pkg: vextest/vec
cpu: Common KVM processor
BenchmarkVecAll/array_any         	1000000000	         0.002960 ns/op
BenchmarkVecAll/vec_any           	1000000000	         0.01795 ns/op
PASS
ok  	vextest/vec	0.137s
```

Reply:
I maybe totally missing something as a go noob